bundlewrap/items/pkg_apt.py

```diff
@@ -14,8 +15,46 @@ class AptPkg(Pkg):
     WHEN_CREATING_ATTRIBUTES = {
         'start_service': True,
     }
+    _pkg_manual_cache = {}
+
+    def cdict(self):
+        return {
+            'installed': self.attributes['installed'],
+            'mark': 'manual' if self.attributes['installed'] else None,
+        }
+
+    def sdict(self):
+        mark = None
+        if self.attributes['installed']:
+            if self.pkg_manually_installed():
+                mark = 'manual'
+            else:
+                mark = 'auto'
+        return {
+            'installed': self.pkg_installed_cached(),
+            'mark': mark,
+        }
+
+    def fix(self, status):
+        with suppress(KeyError):
+            self._pkg_install_cache.get(self.node.name, set()).remove(self.id)
+        pkg_name = self.name.replace("_", ":")
+        if 'installed' in status.keys_to_fix:
+            if self.attributes['installed'] is False:
+                self.pkg_remove()
+            else:
+                self._pkg_manual_cache.setdefault(self.node.name, set()).add(pkg_name)
+                self.pkg_install()
+        elif 'mark' in status.keys_to_fix:
+            self._pkg_manual_cache.setdefault(self.node.name, set()).add(pkg_name)
+            self.run("apt-mark manual {}".format(quote(pkg_name)))
 
     def pkg_all_installed(self):
+        result = self.run("apt-mark showmanual")
```

Comment:
Populating the cache in this function (`pkg_all_installed()`) is racy, sadly. It’s not guaranteed that this function gets called *before* we query the cache for the first time.

Either use the pattern as `pkg_installed_cached()` / `pkg_all_installed()`, or maybe just move populating the cache into `pkg_manually_installed()`?

Reply:
In my tests, this worked every time. I have moved the function as requested nonetheless.